promql/functions.go

```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
+	if f := s[i]; !math.IsNaN(f) {
+		return f
+	}
+
+	x := sf * d[i]
+	y := (1 - sf) * (doubleSVal(i-1, sf, tf, s, b, d) + doubleBVal(i-1, sf, tf, s, b, d))
+
+	// cache value
+	s[i] = x + y
+
+	return s[i]
+}
+
+// Calculate the trend value at the given index i in raw data d.
+// This is somewhat analogous to the slope of the trend at the given index.
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleBVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
+	if f := b[i]; !math.IsNaN(f) {
+		return f
+	}
+
+	x := tf * (doubleSVal(i, sf, tf, s, b, d) - doubleSVal(i-1, sf, tf, s, b, d))
+	y := (1 - tf) * doubleBVal(i-1, sf, tf, s, b, d)
+
+	// cache value
+	b[i] = x + y
+
+	return b[i]
+}
+
+// === holt_winters(node model.ValVector, smoothingFactor model.ValScalar, trendFactor model.ValScalar) Vector ===
+// Holt-Winters is similar to a weighted moving average, where historical data
+// has exponentially less influence on the current data. Holt-Winter also accounts
+// for trends in data. The smoothing factor (0 < sf < 1) effects how historical data will effect the current
+// data. A lower smoothing factor increases the influence of historical data. The trend factor (0 < tf < 1) effects
+// how trends in historical data will effect the current data. A higher trend factor increases the influence
+// of trends.
+// algorithm taken from https://en.wikipedia.org/wiki/Exponential_smoothing titled: "Double exponential smoothing"
+func funcHoltWinters(ev *evaluator, args Expressions) model.Value {
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// sanity check input
+	if sf <= 0 || sf >= 1 {
+		ev.errorf("invalid smoothing factor. Expected: 0 < tf < 1 got: %f", sf)
+	}
+	if tf <= 0 || tf >= 1 {
+		ev.errorf("invalid trend factor. Expected: 0 < tf < 1 got: %f", sf)
+	}
+
+	// make an output vector large enough to hole the entire result
+	resultVector := make(vector, 0, len(mat))
+
+	// create scratch values
+	var s, b, d []float64
+
+	var l int
+	for _, samples := range mat {
+		l = len(samples.Values)
+
+		// can't do the smoothing operation with less than two points
+		if l < 2 {
+			continue
+		}
+
+		// resize scratch values
+		if l != len(s) {
+			s = make([]float64, l)
+			b = make([]float64, l)
+			d = make([]float64, l)
+		}
+
+		// holt-winters is undefined for NaN
+		foundNaN := false
+		// fill in the d values with the raw values from the input
+		for i, v := range samples.Values {
+			d[i] = float64(v.Value)
+
+			// holt-winters is undefined for NaN
+			if math.IsNaN(d[i]) {
+				foundNaN = true
+				break
+			}
+		}
+
+		// skip to next because we found a NaN in the raw data set
```

Comment:
The way we're currently handling it would be to result in that having a NaN output